mantle/kola/harness.go

```diff
@@ -973,20 +973,30 @@ func IsWarningOnFailure(testName string) bool {
 	return false
 }
 
-func getRerunnable(tests []*harness.H) []string {
-	var testsToRerun []string
-	for _, h := range tests {
+func getRerunnable(testsBank map[string]*register.Test, testResults []*harness.H) map[string]*register.Test {
+	// First get the names of the tests that need to rerun
+	var testNamesToRerun []string
+	for _, h := range testResults {
 		// The current nonexclusive test wrapper would have all non-exclusive tests.
 		// We would add all those tests for rerunning if none of the non-exclusive
 		// subtests start due to some initial failure.
 		if nonexclusiveWrapperMatch.MatchString(h.Name()) && !h.GetNonExclusiveTestStarted() {
 			if h.Failed() {
-				testsToRerun = append(testsToRerun, h.Subtests()...)
+				testNamesToRerun = append(testNamesToRerun, h.Subtests()...)
 			}
 		} else {
 			name, isRerunnable := GetRerunnableTestName(h.Name())
 			if h.Failed() && isRerunnable {
-				testsToRerun = append(testsToRerun, name)
+				testNamesToRerun = append(testNamesToRerun, name)
+			}
+		}
+	}
+	// Then convert the list of names into a list of a register.Test objects
+	testsToRerun := make(map[string]*register.Test)
+	for name, t := range testsBank {
+		for _, rerun := range testNamesToRerun {
+			if name == rerun {
```

Comment:
I presume this works as expected for non-exclusive external tests?

Reply:
it seemed to work in a local test :) I had artificially added some tags to some tests and made one of them fail and marked some of them as non-exclusive:

<details>

```
[dustymabe@media fcos]$ cosa kola run --rerun --tag=files --parallel=4                                                                                                                                                                                                                                                     
COREOS_ASSEMBLER_CONTAINER_RUNTIME_ARGS=-v /var/b/shared/code/github.com/coreos/coreos-assembler//bin/:/usr/local/bin/:ro -v /var/b/shared/code/github.com/coreos/coreos-assembler/:/src/:rw                                                                                                                               
COREOS_ASSEMBLER_GIT=/var/b/shared/code/github.com/coreos/coreos-assembler/                                                                                                                                                                                                                                                
COREOS_ASSEMBLER_CONFIG_GIT=/var/b/shared/code/github.com/coreos/fedora-coreos-config/                                                                                                                                                                                                                                     
COREOS_ASSEMBLER_CONTAINER_NAME=cosafcos                                                                                                                                                                                                                                                                                   
+ podman run --rm -ti --security-opt label=disable --privileged --uidmap=1000:0:1 --uidmap=0:1:1000 --uidmap 1001:1001:64536 -v /var/b/shared/assembler/fcos:/srv/ --device /dev/kvm --device /dev/fuse --tmpfs /tmp -v /var/tmp:/var/tmp --name cos4                                                                      
kola -p qemu run --rerun --tag=files --parallel=4 --output-dir tmp/kola                                                                                                                                                                                                                                                    
⏭️  Skipping kola test pattern "fcos.internet":                                                                                                                                                                                                                                                                             
  👉 https://github.com/coreos/coreos-assembler/pull/1478                                                                                                                                                                                                                     
⏭️  Skipping kola test pattern "podman.workflow":                                                                                                                                                                                                                              
  👉 https://github.com/coreos/coreos-assembler/pull/1478                                                                                                                                                                                                                     
=== RUN   ext.config.files.check-symlink                                                                                                                                                                                                                                      
=== RUN   non-exclusive-test-bucket-0                                                                                                                                                                                                                                         
=== RUN   non-exclusive-test-bucket-0/ext.config.files.license                                                                                                                                                                                                                
systemctl status kola-runext-23.service:                                                                                                                                                                                                                                      
× kola-runext-23.service                                                                                                                                                                     
     Loaded: loaded (/etc/systemd/system/kola-runext-23.service; static)                                                                                                                     
    Drop-In: /usr/lib/systemd/system/service.d                                                                                                                                               
             └─10-timeout-abort.conf                                                                                                                                                         
     Active: failed (Result: exit-code) since Tue 2023-09-19 20:28:51 UTC; 100ms ago                                                                                                         
   Duration: 7ms                                                                                                                                                                             
    Process: 2131 ExecStart=/usr/local/bin/kola-runext-license (code=exited, status=1/FAILURE)                                                                                               
   Main PID: 2131 (code=exited, status=1/FAILURE)                                                                                                                                            
        CPU: 3ms                                                                                                                                                                             
                                                                                                                                                                                             
Sep 19 20:28:51 qemu0 systemd[1]: Started kola-runext-23.service.                                                                                                                            
Sep 19 20:28:51 qemu0 kola-runext-license[2131]: + test -f /usr/share/licenses/fedora-coreos-config/LICENSEE                                                                                 
Sep 19 20:28:51 qemu0 kola-runext-license[2131]: + fatal missing LICENSE                                                                                                                     
Sep 19 20:28:51 qemu0 kola-runext-license[2131]: + echo missing LICENSE                                                                                                                      
Sep 19 20:28:51 qemu0 kola-runext-license[2131]: missing LICENSE                                                                                                                             
Sep 19 20:28:51 qemu0 kola-runext-license[2131]: + exit 1                     
Sep 19 20:28:51 qemu0 systemd[1]: kola-runext-23.service: Main process exited, code=exited, status=1/FAILURE                                                                                                                                                                                                               
Sep 19 20:28:51 qemu0 systemd[1]: kola-runext-23.service: Failed with result 'exit-code'.                                                                    
=== RUN   non-exclusive-test-bucket-0/ext.config.files.aleph-version                                                                                         
--- PASS: ext.config.files.check-symlink (26.99s)                             
--- FAIL: non-exclusive-test-bucket-0 (27.59s)                                
    --- FAIL: non-exclusive-test-bucket-0/ext.config.files.license (0.55s)                                                                                   
            cluster.go:162: Error: Unit kola-runext-23.service exited with code 1                                                                            
            cluster.go:162: 2023-09-19T20:28:51Z cli: Unit kola-runext-23.service exited with code 1                                                                                                                                                                                                                       
            harness.go:1262: kolet failed: : kolet run-test-unit failed: Process exited with status 1                                                                                                                                                                                                                      
    --- PASS: non-exclusive-test-bucket-0/ext.config.files.aleph-version (1.50s)                                                                             
FAIL, output in tmp/kola                                                      


======== Re-running failed tests (flake detection) ========                   

⏭️  Skipping kola test pattern "fcos.internet":                                
  👉 https://github.com/coreos/coreos-assembler/pull/1478                     
⏭️  Skipping kola test pattern "podman.workflow":                              
  👉 https://github.com/coreos/coreos-assembler/pull/1478                     
=== RUN   ext.config.files.license                                            
systemctl status kola-runext-23.service:                                      
× kola-runext-23.service                                                      
     Loaded: loaded (/etc/systemd/system/kola-runext-23.service; static)                                                                                     
    Drop-In: /usr/lib/systemd/system/service.d                                
             └─10-timeout-abort.conf                                          
     Active: failed (Result: exit-code) since Tue 2023-09-19 20:29:19 UTC; 112ms ago                                                                         
   Duration: 6ms                                                              
    Process: 2078 ExecStart=/usr/local/bin/kola-runext-license (code=exited, status=1/FAILURE)                                                                                                                                                                                                                             
   Main PID: 2078 (code=exited, status=1/FAILURE)                             
        CPU: 2ms                                                              

Sep 19 20:29:19 qemu0 systemd[1]: Started kola-runext-23.service.                                                                                            
Sep 19 20:29:19 qemu0 kola-runext-license[2078]: + test -f /usr/share/licenses/fedora-coreos-config/LICENSEE                                                                                                                                                                                                               
Sep 19 20:29:19 qemu0 kola-runext-license[2078]: + fatal missing LICENSE                                                                                     
Sep 19 20:29:19 qemu0 kola-runext-license[2078]: + echo missing LICENSE                                                                                      
Sep 19 20:29:19 qemu0 kola-runext-license[2078]: missing LICENSE                                                                                             
Sep 19 20:29:19 qemu0 kola-runext-license[2078]: + exit 1                     
Sep 19 20:29:19 qemu0 systemd[1]: kola-runext-23.service: Main process exited, code=exited, status=1/FAILURE                                                                                                                                                                                                               
Sep 19 20:29:19 qemu0 systemd[1]: kola-runext-23.service: Failed with result 'exit-code'.                                                                    
--- FAIL: ext.config.files.license (26.58s)                                   
        cluster.go:162: Error: Unit kola-runext-23.service exited with code 1                                                                                
        cluster.go:162: 2023-09-19T20:29:19Z cli: Unit kola-runext-23.service exited with code 1                                                                                                                                                                                                                           
        harness.go:1262: kolet failed: : kolet run-test-unit failed: Process exited with status 1                                                                                                                                                                                                                          
FAIL, output in tmp/kola/rerun                                                
Error: harness: test suite failed                                             
2023-09-19T20:29:22Z cli: harness: test suite failed                          
failed to execute cmd-kola: exit status 1
```

</details>